File: console/src/api/modules/monitor.ts
/**
 * 监控 API 模块
 * v4.0 新增 - 7 个 API 接口，与后端 routes/monitor.py 完全匹配
 */
import { request } from "../request";

/**
 * 指标接口
 */
export interface MonitorMetric {
  metric_id: string;
  metric_name: string;
  metric_type: string;
  value: number;
  unit: string;
  timestamp: string;
  labels?: Record<string, string>;
}

/**
 * 告警接口
 */
export interface MonitorAlert {
  alert_id: string;
  alert_name: string;
  metric_id: string;
  threshold: number;
  operator: string;
  severity: "INFO" | "WARNING" | "ERROR" | "CRITICAL";
  status: "active" | "resolved" | "muted";
  channels: string[];
  created_at: string;
  triggered_at?: string;
  resolved_at?: string;
}

/**
 * 告警创建请求
 */
export interface AlertCreateRequest {
  alert_name: string;
  metric_id: string;
  threshold: number;
  operator: string;
  severity: "INFO" | "WARNING" | "ERROR" | "CRITICAL";
  channels: string[];
}

/**
 * 告警更新请求
 */
export interface AlertUpdateRequest {
  alert_name?: string;
  threshold?: number;
  operator?: string;
  severity?: "INFO" | "WARNING" | "ERROR" | "CRITICAL";
  channels?: string[];
}

/**
 * 获取系统指标
 * @param tenantId 租户 ID（可选）
 * @returns 指标列表
 */
export async function getMetrics(tenantId?: string): Promise<MonitorMetric[]> {
  const params = new URLSearchParams();
  if (tenantId) {
    params.append("tenant_id", tenantId);
  }
  const query = params.toString() ? `?${params.toString()}` : "";
  return request(`/monitor/metrics${query}`);
}

/**
 * 获取指标详情
 * @param metricId 指标 ID
 * @returns 指标详情
 */
export async function getMetric(metricId: string): Promise<MonitorMetric> {
  return request(`/monitor/metrics/${metricId}`);
}

/**
 * 列出告警
 * @param tenantId 租户 ID（可选）
 * @param status 告警状态（可选）
 * @returns 告警列表
 */
export async function listAlerts(
  tenantId?: string,
  status?: string
): Promise<MonitorAlert[]> {
  const params = new URLSearchParams();
  if (tenantId) {
    params.append("tenant_id", tenantId);
  }
  if (status) {
    params.append("status", status);
  }
  const query = params.toString() ? `?${params.toString()}` : "";
  return request(`/monitor/alerts${query}`);
}

/**
 * 获取告警详情
 * @param alertId 告警 ID
 * @returns 告警详情
 */
export async function getAlert(alertId: string): Promise<MonitorAlert> {
  return request(`/monitor/alerts/${alertId}`);
}

/**
 * 创建告警
 * @param data 告警创建请求
 * @returns 创建后的告警信息
 */
export async function createAlert(data: AlertCreateRequest): Promise<MonitorAlert> {
  return request("/monitor/alerts", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 更新告警
 * @param alertId 告警 ID
 * @param data 告警更新请求
 * @returns 更新后的告警信息
 */
export async function updateAlert(
  alertId: string,
  data: AlertUpdateRequest
): Promise<MonitorAlert> {
  return request(`/monitor/alerts/${alertId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 删除告警
 * @param alertId 告警 ID
 * @returns 操作结果
 */
export async function deleteAlert(alertId: string): Promise<{ message: string }> {
  return request(`/monitor/alerts/${alertId}`, {
    method: "DELETE",
  });
}

/**
 * 监控 API 模块导出
 */
export const monitorApi = {
  getMetrics,
  getMetric,
  listAlerts,
  getAlert,
  createAlert,
  updateAlert,
  deleteAlert,
};

File: src/copaw/app/routers/monitor.py
# -*- coding: utf-8 -*-
"""Monitor API routes."""
from __future__ import annotations

import logging
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/monitor", tags=["monitor"])


METRICS_DB: Dict[str, Dict[str, Any]] = {}
ALERTS_DB: Dict[str, Dict[str, Any]] = {}
ALERT_HISTORY: List[Dict[str, Any]] = []


class MonitorMetric(BaseModel):
    """Monitor metric response."""

    metric_id: str
    metric_name: str
    metric_type: Literal["gauge", "counter", "histogram"]
    value: float
    unit: str
    timestamp: str
    labels: Optional[Dict[str, str]] = None


class MonitorAlert(BaseModel):
    """Monitor alert response."""

    alert_id: str
    alert_name: str
    metric_id: str
    threshold: float
    operator: Literal["gt", "lt", "gte", "lte", "eq"]
    severity: Literal["INFO", "WARNING", "ERROR", "CRITICAL"]
    status: Literal["active", "resolved", "muted"]
    channels: List[str]
    created_at: str
    triggered_at: Optional[str] = None
    resolved_at: Optional[str] = None


class AlertCreateRequest(BaseModel):
    """Alert create request."""

    alert_name: str
    metric_id: str
    threshold: float
    operator: Literal["gt", "lt", "gte", "lte", "eq"]
    severity: Literal["INFO", "WARNING", "ERROR", "CRITICAL"]
    channels: List[str]


class AlertUpdateRequest(BaseModel):
    """Alert update request."""

    alert_name: Optional[str] = None
    threshold: Optional[float] = None
    operator: Optional[Literal["gt", "lt", "gte", "lte", "eq"]] = None
    severity: Optional[Literal["INFO", "WARNING", "ERROR", "CRITICAL"]] = None
    channels: Optional[List[str]] = None
    status: Optional[Literal["active", "resolved", "muted"]] = None


def _get_current_user(request: Request) -> Dict[str, str]:
    """Get current user from request state."""
    user = getattr(request.state, "user", None)
    if not user:
        return {
            "user_id": "default",
            "tenant_id": "default",
            "role": "tenant_admin",
        }
    return {
        "user_id": user,
        "tenant_id": "default",
        "role": "tenant_admin",
    }


@router.get("/metrics", response_model=List[MonitorMetric])
async def get_metrics(
    request: Request,
    tenant_id: Optional[str] = Query(None),
) -> List[Dict[str, Any]]:
    """Get system metrics."""
    _get_current_user(request)
    return list(METRICS_DB.values())


@router.get("/metrics/{metric_id}", response_model=MonitorMetric)
async def get_metric(
    request: Request,
    metric_id: str,
) -> Dict[str, Any]:
    """Get metric details."""
    _get_current_user(request)
    metric = METRICS_DB.get(metric_id)
    if not metric:
        raise HTTPException(
            status_code=404,
            detail=f"Metric not found: {metric_id}",
        )
    return metric


@router.get("/alerts", response_model=List[MonitorAlert])
async def list_alerts(
    request: Request,
    tenant_id: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
) -> List[Dict[str, Any]]:
    """List alerts."""
    _get_current_user(request)
    alerts = list(ALERTS_DB.values())
    if status:
        alerts = [a for a in alerts if a.get("status") == status]
    return alerts


@router.get("/alerts/{alert_id}", response_model=MonitorAlert)
async def get_alert(
    request: Request,
    alert_id: str,
) -> Dict[str, Any]:
    """Get alert details."""
    _get_current_user(request)
    alert = ALERTS_DB.get(alert_id)
    if not alert:
        raise HTTPException(
            status_code=404,
            detail=f"Alert not found: {alert_id}",
        )
    return alert


@router.post("/alerts", response_model=MonitorAlert)
async def create_alert(
    request: Request,
    data: AlertCreateRequest,
) -> Dict[str, Any]:
    """Create alert."""
    _get_current_user(request)
    alert_id = f"alert_{data.alert_name}_{int(datetime.utcnow().timestamp())}"
    alert = {
        "alert_id": alert_id,
        "alert_name": data.alert_name,
        "metric_id": data.metric_id,
        "threshold": data.threshold,
        "operator": data.operator,
        "severity": data.severity,
        "status": "active",
        "channels": data.channels,
        "created_at": datetime.utcnow().isoformat() + "Z",
        "triggered_at": None,
        "resolved_at": None,
    }
    ALERTS_DB[alert_id] = alert
    return alert


@router.put("/alerts/{alert_id}", response_model=MonitorAlert)
async def update_alert(
    request: Request,
    alert_id: str,
    data: AlertUpdateRequest,
) -> Dict[str, Any]:
    """Update alert."""
    _get_current_user(request)
    alert = ALERTS_DB.get(alert_id)
    if not alert:
        raise HTTPException(
            status_code=404,
            detail=f"Alert not found: {alert_id}",
        )
    if data.alert_name:
        alert["alert_name"] = data.alert_name
    if data.threshold is not None:
        alert["threshold"] = data.threshold
    if data.operator:
        alert["operator"] = data.operator
    if data.severity:
        alert["severity"] = data.severity
    if data.channels:
        alert["channels"] = data.channels
    if data.status:
        alert["status"] = data.status
        if data.status == "resolved":
            alert["resolved_at"] = datetime.utcnow().isoformat() + "Z"
    return alert


@router.delete("/alerts/{alert_id}")
async def delete_alert(
    request: Request,
    alert_id: str,
) -> Dict[str, str]:
    """Delete alert."""
    _get_current_user(request)
    if alert_id not in ALERTS_DB:
        raise HTTPException(
            status_code=404,
            detail=f"Alert not found: {alert_id}",
        )
    del ALERTS_DB[alert_id]
    return {"message": "Alert deleted successfully"}
